bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java

```diff
@@ -0,0 +1,173 @@
+
+public class ReportDetailView extends ReportDetailViewBase implements Receiver, SucceededListener, StartedListener, ProgressListener{
+
+
+	private ReportsModel model;
+	private Binder<Report> binder = new Binder<Report>();
+	
+	public ReportDetailView(ReportsModel reportModel) {
+		model = reportModel;
+		initializeBinder();
+		initializeUIComponents();
+		init();
+	}
+	
+
+	private void initializeBinder() {
+		binder.bind(cmbEditPriority, Report::getPriority, Report::setPriority);
+		binder.bind(cmbEditType, Report::getType, Report::setType);
+		binder.bind(cmbEditStatus, Report::getStatus, Report::setStatus);
+		binder.bind(cmbEditAssignedTo, Report::getAssigned, Report::setAssigned);
+		binder.bind(cmbEditVersion, Report::getVersion, Report::setVersion);
+	}
+
+	private void initializeUIComponents() {
+		btnUpdateReport.addClickListener(evt -> saveReport());
+		btnRevertReport.addClickListener(evt -> revertChanges());
+		btnDone.addClickListener(evt -> saveComment());
+		btnCancel.addClickListener(evt -> closeWindow());
+		
+		btnUpload.setReceiver(this);
+		btnUpload.addStartedListener(this);
+		btnUpload.addSucceededListener(this);
+		btnUpload.addProgressListener(this);
+	}
+
+	private void init() {
+		((HorizontalLayout)pnlAttachments.getContent()).removeAllComponents();
+		pnlAttachments.setVisible(false);
+		Report report = model.getReportForEdit();
+		binder.setBean(model.getReportForEdit());
+		lblProjectSummary.setValue(report.getSummary());
+		initializeComboContents();
+		fillComments();
+	}
+	
+	private void initializeComboContents() {
+		cmbEditPriority.setItems(model.getPriorties());
+		cmbEditType.setItems(model.getTypes());
+		cmbEditStatus.setItems(model.getStatuses());
+		cmbEditAssignedTo.setItems(model.findReporters());
+		cmbEditVersion.setItems(model.findProjectVersions(model.getReportForEdit().getProject()));
+	}
+	
+	private void fillComments() {
+		VerticalLayout layout = (VerticalLayout) pnlCommentsThread.getContent();
+		layout.removeAllComponents();
+		List<Comment> comments = model.getComments();
+		for (Comment comment : comments) {
+			layout.addComponent(new ThreadItemView(new ThreadItemModel(comment)));
+		}
+	}
+	
+	
+	private void saveReport() {
+		model.save();
+		Notification.show("Saved successfully", Type.TRAY_NOTIFICATION);
+	}
+	
+	private void revertChanges() {
+		model.resetReportForEdit();
+		binder.setBean(model.getReportForEdit());
+	}
+
+	private void saveComment() {
+		if (!txtComment.getOptionalValue().isPresent())
+			return;
+		model.saveComment(txtComment.getValue(), BaseModel.loginUser);
+		fillComments();
+		txtComment.clear();
+	}
+
+
+	@Override
+	public void uploadStarted(StartedEvent event) {
+		pnlAttachments.setVisible(true);
+		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
+		ProgressBar progress = new ProgressBar();
+		progress.setCaption(event.getFilename());
+		model.getAttachmentUIElements().put(event.getFilename(), progress);
```

Comment:
What happens if two attachments have the same filename? Is there something else you could be using as the map key?